[UI] New volumes page - Metrics & Details Tabs

alexis-ld opened the pull request:
**Component**:

ui, volumes

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

We want to adapt the volumes page so it looks like the new node page (tab based).

**This specific PR adds the Metrics and Details tabs as well as the new table styling. It is meant to be merged in the feature branch #2835** 

**Summary**:

I refactored most of the Volume view. There were also small adaptations compared to the base view (alerts filtering in URL, ...) as described in #2751 .

Some components (ActiveAlertsFilters, ActiveAlertsCounter) were designed to be re-usable in other views.

![Screenshot 2020-10-15 at 15 41 13](https://user-images.githubusercontent.com/3480526/96138852-7dccfc00-0efe-11eb-8069-0c2f22e0b8f9.png)
![Screenshot 2020-10-15 at 15 41 26](https://user-images.githubusercontent.com/3480526/96138858-80c7ec80-0efe-11eb-9b3a-935befdb6a77.png)
![Screenshot 2020-10-15 at 15 41 46](https://user-images.githubusercontent.com/3480526/96138868-832a4680-0efe-11eb-84e7-e865f05b759d.png)
![Screenshot 2020-10-15 at 15 41 59](https://user-images.githubusercontent.com/3480526/96138875-84f40a00-0efe-11eb-92e3-13dfacacf558.png)



**Acceptance criteria**: 

* Volumes page right panel is separated in four tabs Overview, Alerts, Metrics, Details
* Total number of alerts is displayed in Alerts tab header
* It is possible to filter by alert severity on the Alert tab
* Clicking on an alert category in the overview tab opens the alerts tab and filters by the corresponding category
* When switching to another volume the selected tab is preserved
---


Refs: #2751

Commit:
ui/volumes: Add translations for new volume page
Refs: #2751

Commit:
ui/volumes:
Add optional parameter to /volumes/ route to use volume name more easily in sub components
Refs: #2751

Commit:
ui/volumes:
Re-usable components to filter/count alerts for a volume
Refs: #2751

Commit:
ui/volumes:
TextBadge component to display a counter in a block

Commit:
ui/volumes:
Refactor VolumeAlertsCard component to serve as a tab in new Volumes view
Refs: #2751

Commit:
ui/volumes:
Refactors VolumeDetailsCard as the overview tab for the new volumes page
Refs: #2751

Commit:
ui/volumes:
Adapts Volume table to forward URLs properly to keep tabs selection on the right side.
Refs: #2751

Commit:
ui/volumes:
Add tab-based right panel in on Volumes page
Refs: #2751

Commit:
ui/volumes:
Use Switch from react-router-dom in tabs to display selected content to match coding style.

Refs: #2751

Commit:
ui/volumes:
Re-enable metrics component to use it as a tab in the new Volume tab-based view.

Refs: #2751

Commit:
ui/volumes:
Improve Charts layout to take advantage of the available space in the tab.
Add Advanced Metrics button.

Refs: #2751

Commit:
ui/volumes:
Implement Redux flow to retrieve a volume object.

Commit:
ui/volumes:
Add details tab to display the selected volume object in the new volume page.

Refs: #2751

Commit:
ui/volumes:
Adapt colors based on the new theme variables to fit to the design

Refs: #2751

Commit:
ui/volumes:
Make node name clickable in overview tab. On click it redirects to the corresponding node page.

Refs: #2751

Commit:
ui/volumes:
Adapt Metrics tab styling after theme update to fit to the design.

Refs: #2751

Commit:
ui/volumes:
Adapt table to fit the new design.

Refs: #2751

thomasdanan commented:
Hello @alexis-ld 

Thanks for the PR, it is pretty cool.
Just few stuffs I noticed while testing quickly:

- When clicking on Advanced Metrics, I always land on the loki-0 persistent Volume. Probably something missing in the URL
- I should see one Critical alert for Volume master-0-alertmanager but I don't see it
- The scrolling is global to the right-hand side of the page but it would be better if it was within the tab to always see the list of the available tab (see tab detail)
- Also when landing on Volume page, it would be good to have the first Volume selected and related info displayed in the right-hand side (I am not fan of seeing an empty right-hand side panel). Actually, Volume should be sorted by Health severity and if there is a Volume with Critical Health, I would like to see information about this one.

Thanks

Thomas

Commit:
ui/volumes:
Make tabs scrollable with a max-height to scroll within the tab instead of the page.

Refs: #2751

Commit:
ui/volumes:
Improve volume listing. Sort by health (unhealthy first)

Commit:
ui/volumes:
Auto selects first volume on Volumes page to avoid empty right side.

Refs: #2751

alexis-ld commented:
@thomasdanan 

I just push a few changes.

- The tabs now have a max-width depending on the view height and are scrollable as requested
- The volumes list now have an initial sort by health
- If no volumes is selected the first (most unhealthy volume) is selected by default

ChengYanJin commented on ui/src/components/VolumeMetricsTab.js:
```diff
@@ -33,20 +33,26 @@ import { intl } from '../translations/IntlGlobalProvider';
-  background-color: ${(props) => props.theme.brand.primaryDark1};
-  margin: ${padding.small};
-  padding-bottom: ${padding.large};
+
+  .sc-vegachart svg {
+    background-color: inherit !important;
+  }
```
I would suggest obtaining the interior width/height of the window, and compute the size of th LineChart
We can do something like this:
```javascipt
<LineChart
  id,
  data,
  xAxis,
  yAxis,
  color,
  width={window.innerWidth / 4 - 50}
  height={window.innerHeight / 6 - 30}
/>
```

ChengYanJin commented on ui/src/containers/VolumePage.js:
```diff
@@ -14,6 +15,8 @@ import {
   stopRefreshPersistentVolumesAction,
   fetchPersistentVolumeClaimAction,
+  fetchCurrentVolumeObjectAction,
+  clearCurrentVolumeObjectAction,
```
Not sure we need `clearCurrentVolumeObjectAction`. IMO, we can keep the volume object of the last selected volume in the store until we select another volume.

alexis-ld replied:
IMO the clear is useful so we don't display the previous volume data while the new request is pending.
It that is not an issue I can remove the clear action. Let me know :)

alexis-ld replied:
Removed it :)

ChengYanJin commented on ui/src/ducks/app/volumes.js:
```diff
@@ -43,6 +46,10 @@ const defaultState = {
   isLoading: false,
   isSCLoading: false,
+  currentVolumeObject: {
+    error: null,
```
IMO, we don't need to store the `error`  since we don't want to display anything from error in UI.

alexis-ld replied:
Ok, I thought some error handling on the UI was needed :) removing.

ChengYanJin commented on ui/src/components/VolumeDetailsTab.js:
```diff
@@ -0,0 +1,35 @@
+const DetailsTabContainer = styled.div`
+  color: ${(props) => props.theme.brand.textPrimary};
+`;
+
```
We have `TabContainer` in CommonLayoutStyle.js 
Please update the style there and apply this for all the tabs containers.

ChengYanJin commented on ui/src/components/VolumeDetailsTab.js:
```diff
@@ -0,0 +1,35 @@
+        {data && JSON.stringify(data, null, '\t')}
+        {error && (
+          <ErrorText>{intl.translate('error_volume_details')}</ErrorText>
+        )}
```
If order to check if there is an error, you can check if data is `null`.
Thanks to the job you did when fetch the current volume object.

```javascript
export function* fetchCurrentVolumeObject({ volumeName }) {
  const result = yield call(VolumesApi.getVolumeObject, volumeName);
  if (!result.error) {
    yield put(setCurrentVolumeObjectAction({ data: result.body, error: null }));
  } else {
    yield put(
      setCurrentVolumeObjectAction({ data: null, error: result.error }),
    );
  }
}
```

ChengYanJin commented on ui/src/components/VolumeDetailsTab.js:
```diff
@@ -0,0 +1,35 @@
+`;
+
+const VolumeObjectContent = styled.div`
+  white-space: pre-wrap;
```
the font-size of the VolumeObject should be **14px** acoording to the design ;)

ChengYanJin commented on ui/src/containers/VolumePage.js:
```diff
@@ -101,6 +104,16 @@ const VolumePage = (props) => {
+  useEffect(() => {
+    if (volumeListData.length && !currentVolumeName) {
+      history.push({
+        pathname: `/volumes/${volumeListData[0]?.name}/overview`,
```
```suggestion
        pathname: `/volumes/${currentVolumeName}/overview`,
```

alexis-ld replied:
Not sure I understand there. We want to push to history when there is **no** currentVolumeName. In that condition currentVolumeName is undefined.

ChengYanJin replied:
Sorry. My bad. You're correct here!

ChengYanJin commented on ui/src/containers/VolumePage.js:
```diff
@@ -101,6 +104,16 @@ const VolumePage = (props) => {
     getVolumeListData(state, props),
   );
 
+  // If data has been retrieved and no volume is selected yet we select the first one
```
Not sure we want to auto-select the first volume when we arrive `/volumes`. I am wondering is this what you want to achieve?

ChengYanJin reviewed (changes requested):
Hello @alexis-ld ,

Mainly some comments about redux-saga to retrieve the volume object. Please come back to me if you have any questions. 

Thank you!

Commit:
ui/volumes:
Change charts height and width in Metrics tab so they are dynamic.

Refs: #2751

Commit:
ui/volumes:
Remove clear volume object action, we don't need it since the API response on this endpoint is very fast.

Commit:
ui/volumes:
Remove error from Volume object data structure as we don't want to display error messages in the UI directly.

Commit:
ui/volumes:
Set details tab font-size to fontSize.base to fit the design.

Refs: #2751

Commit:
ui/volumes:
Refactor ActiveAlertsCounter component to pass counters directly from props.

Commit:
ui/volumes:
Move URL handling directly into ActiveAlertsFilter component.

Commit:
ui/volumes: Remove &nbsp's and use margin instead.

Commit:
ui/volumes:
Update color code in ActiveAlertsCounter to fit design.

Commit:
ui/volumes:
Remove condition for overview tab selection as /overview is now default in the URL.

Refs: #2751

Commit:
ui/volumes:
Remove &nbsp; and use padding instead.

Commit:
ui/volumes: Removed fa-2x property.

Commit:
ui/volumes:
Remove baselink prop from ActiveAlertsCounter component handling the URL in the component directly.

Commit:
ui/volumes:
Changed tab styling to match the new design. Improved contrast and readability.

Refs: #2751

Commit:
ui/volumes:
Refactor tab to avoid label wrapping on small widths.

alexis-ld commented:
Made all the changes, rebased from `improvement/2751-new-volumes-page`, fixed the conflicts.

thomasdanan reviewed (dismissed):
thanks @alexis-ld this is perfect.

I just noticed one little bug when you click on create a volume and then Cancel, you are not redirected to the previous page

ChengYanJin commented:
Hi @alexis-ld ,
For the issue that Thomas raised, maybe you want to fix the Cancel button like this?
```javascript               
<Button
  text={intl.translate('cancel')}
  type="button"
  outlined
  onClick={() => history.goBack()}
/>
```

Commit:
ui/volumes:
Fix Cancel button on create volume view not redirecting to the new volume page

alexis-ld commented:
@ChengYanJin Thank you very much for the suggestion :) 
@thomasdanan Just pushed the fix.

ChengYanJin reviewed (dismissed):
LGTM!

alexis-ld commented:
/approve

bert-e commented:
# Hello alexis-ld,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2751-new-volumes-page-metrics-details` into
`development/2.6`
* [w/2.7/improvement/2751-new-volumes-page-metrics-details](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2751-new-volumes-page-metrics-details) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal










*The following options are set:* **approve**

ChengYanJin reviewed (dismissed):
LGTM!

alexis-ld commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/7d3fef55f2c7bbdc64efd00bc1aab218d16bb888) did not succeed in branch w/2.7/improvement/2751-new-volumes-page-metrics-details.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/a93d78f73aa25ab66e959a9e1d50e6462820913f) did not succeed in branch improvement/2751-new-volumes-page-metrics-details.




*The following options are set:* **approve**

Commit:
ui/volumes:
Adapt E2E test to new volume page.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal










*The following options are set:* **approve**

ChengYanJin committed:
ui/tests: Adapt the volumes.test

Since we refactored the Volume Page,
after creating the volume, we go to `/volumes/${newVolume.name}/overview?node=${newVolume.node}`

So we adapt the create volume duck and test too

Refs: #2751

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/ee0cd5f67c77015d7251452ce4777d9bccb0861a) did not succeed in branch w/2.7/improvement/2751-new-volumes-page-metrics-details.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/2892ccea96ce45c5c6161af4d37547ae4f714ae4) did not succeed in branch improvement/2751-new-volumes-page-metrics-details.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/ee0cd5f67c77015d7251452ce4777d9bccb0861a) did not succeed in branch w/2.7/improvement/2751-new-volumes-page-metrics-details.




*The following options are set:* **approve**

Commit:
ui/volumes:
Fix duplicated items and missing prop after rebase.

bert-e commented:
# History mismatch



Merge commit #a3dfde57e0bfd64b9263bb0ec584addb7f0289f2 on the integration branch
`w/2.7/improvement/2751-new-volumes-page-metrics-details` is merging a branch which is neither the current
branch `improvement/2751-new-volumes-page-metrics-details` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2751-new-volumes-page-metrics-details` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

alexis-ld commented:
/reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**





*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #a3dfde57e0bfd64b9263bb0ec584addb7f0289f2 on the integration branch
`w/2.7/improvement/2751-new-volumes-page-metrics-details` is merging a branch which is neither the current
branch `improvement/2751-new-volumes-page-metrics-details` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2751-new-volumes-page-metrics-details` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

ChengYanJin committed:
ui/navigation: Fix the bug to navigate from Node page volume tab to Volume Page

Since we redesign the volume page with tabs

Refs: #2751

ChengYanJin committed:
ui/nodes: Fix the fontsize in Node Detail tab to 14px

Refs: #2866

ChengYanJin committed:
ui/styles: Adapt the same css style in Volume and Node Page

Refs: #2866

ChengYanJin commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2751-new-volumes-page-metrics-details` into
`development/2.6`
* [w/2.7/improvement/2751-new-volumes-page-metrics-details](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2751-new-volumes-page-metrics-details) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @alexandre-allard-scality

* @jbertran

* @sayf-eddine-scality

* @teddyandrieux

* @nicolast

* @lucieleonard

* @nootal

* @monpote

* @cmonfort

* @slaperche-scality

* @tcarmet

* @carlito-scality

* @thomasdanan










*The following options are set:* **approve**

ChengYanJin reviewed (approved):
LGTM! Thanks for addressing the conflicts.
BTW, I added some changes/bugfix.

alexis-ld commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/bd70763ae8cd857f1e29b3474cafeeb7514ab531) did not succeed in branch w/2.7/improvement/2751-new-volumes-page-metrics-details.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye alexis-ld.
